charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,252 @@
-# Copyright 2021 Canonical Ltd.
+# Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-import logging
+"""A Juju Charm for Tensorboards Web App."""
 
-from oci_image import OCIImageResource, OCIImageResourceError
+import logging
+from typing import Dict
+
+from charmed_kubeflow_chisme.exceptions import ErrorWithStatus, GenericCharmRuntimeError
+from charmed_kubeflow_chisme.kubernetes import KubernetesResourceHandler
+from charmed_kubeflow_chisme.lightkube.batch import delete_many
+from charmed_kubeflow_chisme.pebble import update_layer
+from charms.observability_libs.v1.kubernetes_service_patch import KubernetesServicePatch
+from lightkube import ApiError
+from lightkube.generic_resource import load_in_cluster_generic_resources
+from lightkube.models.core_v1 import ServicePort
 from ops.charm import CharmBase
 from ops.main import main
-from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+from ops.model import ActiveStatus, BlockedStatus, Container, MaintenanceStatus, WaitingStatus
+from ops.pebble import Layer
 from serialized_data_interface import NoCompatibleVersions, NoVersionsListed, get_interfaces
 
+K8S_RESOURCE_FILES = ["src/templates/auth_manifests.yaml.j2"]
+LIGHTKUBE_FIELD_MANAGER = "lightkube"
 
-class CheckFailed(Exception):
-    """Raise this exception if one of the checks in main fails."""
-
-    def __init__(self, msg: str, status_type=None):
-        super().__init__()
-
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
 
-
-class Operator(CharmBase):
+class TensorboardsWebApp(CharmBase):
     def __init__(self, *args):
+        """Initialize charm and setup the container."""
         super().__init__(*args)
 
-        self.log = logging.getLogger(__name__)
-        self.image = OCIImageResource(self, "oci-image")
-
-        for event in [
-            self.on.install,
-            self.on.upgrade_charm,
-            self.on.config_changed,
-            self.on.leader_elected,
-            self.on["ingress"].relation_changed,
-        ]:
-            self.framework.observe(event, self.main)
+        self.logger = logging.getLogger(__name__)
+        self._namespace = self.model.name
+        self._name = self.app.name
+        self._http_port = self.model.config["port"]
 
-    def main(self, event):
-        try:
+        # Explicitly use defined string since the charm can be deployed with different app name
+        self._container_name = "tensorboards-web-app"
+        self._container = self.unit.get_container(self._container_name)
+        self._k8s_resource_handler = None
 
-            self._check_leader()
-
-            interfaces = self._get_interfaces()
-
-            image_details = self._check_image_details()
+        http_port = ServicePort(int(self._http_port), name="http")
+        self.service_patcher = KubernetesServicePatch(
+            self, [http_port], service_name=f"{self._name}"
+        )
```

Comment:
Continuing the discussion on dropping the port from the config, this is neither useful nor working as expected. For more information you could check out the following issue:
* https://github.com/canonical/oidc-gatekeeper-operator/issues/83